### ISSUE TITLE
Exception thrown when using dynamic table class name and a child component

### ISSUE THREAD

**Issue by Pephers** [lit/lit#3602]
### Which package(s) are affected?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

When I have a web component inside a `table`, an exception is thrown when I either:
1. Use a `class` attribute containing a reference to a variable on the `table`.
2. Use a child web component with a property.

The following code results in an exception being thrown:
```ts
import { html, LitElement } from "lit";
import { customElement, property } from "lit/decorators.js";

@customElement("child-component")
export class ChildComponent extends LitElement {
  @property()
  myProperty = "";

  render() {
    return html`<div>${this.myProperty}</div>`;
  }
}

@customElement("parent-component")
export class ParentComponent extends LitElement {
  render() {
    const className = "my-class";

    return html`<table class=${className}>
      <child-component .myProperty=${"my value"}></child-component>
    </table>`;
  }
}
```
Exception:
```
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'split')
    at new C (lit-html.ts:332:30)
    at ls._$AC (lit-html.ts:1558:53)
    at ls.$ (lit-html.ts:1511:16)
    at ls._$AI (lit-html.ts:1375:12)
    at Z (lit-html.ts:2169:8)
    at xe.update (lit-element.ts:165:24)
    at xe.performUpdate (reactive-element.ts:1331:14)
    at xe.scheduleUpdate (reactive-element.ts:1263:17)
    at xe._$Ej (reactive-element.ts:1235:25)
```

Changing the `table` to a `div` will work as does replacing the `class` attribute with a string like so:

```ts
// Imports and the child component omitted for brevity.

@customElement("parent-component")
export class ParentComponent extends LitElement {
  render() {
    return html`<table class="my-class">
      <child-component .myProperty=${"my value"}></child-component>
    </table>`;
  }
}
```

### Reproduction

To reproduce, try out it out on the Playground:

https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgeyBodG1sLCBMaXRFbGVtZW50IH0gZnJvbSBcImxpdFwiO1xuaW1wb3J0IHsgY3VzdG9tRWxlbWVudCwgcHJvcGVydHkgfSBmcm9tIFwibGl0L2RlY29yYXRvcnMuanNcIjtcblxuQGN1c3RvbUVsZW1lbnQoXCJjaGlsZC1jb21wb25lbnRcIilcbmV4cG9ydCBjbGFzcyBDaGlsZENvbXBvbmVudCBleHRlbmRzIExpdEVsZW1lbnQge1xuICBAcHJvcGVydHkoKVxuICBteVByb3BlcnR5ID0gXCJcIjtcblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPGRpdj4ke3RoaXMubXlQcm9wZXJ0eX08L2Rpdj5gO1xuICB9XG59XG5cbkBjdXN0b21FbGVtZW50KFwicGFyZW50LWNvbXBvbmVudFwiKVxuZXhwb3J0IGNsYXNzIFBhcmVudENvbXBvbmVudCBleHRlbmRzIExpdEVsZW1lbnQge1xuICByZW5kZXIoKSB7XG4gICAgY29uc3QgY2xhc3NOYW1lID0gXCJteS1jbGFzc1wiO1xuXG4gICAgcmV0dXJuIGh0bWxgPHRhYmxlIGNsYXNzPSR7Y2xhc3NOYW1lfT5cbiAgICAgIDxjaGlsZC1jb21wb25lbnQgLm15UHJvcGVydHk9JHtcIm15IHZhbHVlXCJ9PjwvY2hpbGQtY29tcG9uZW50PlxuICAgIDwvdGFibGU-YDtcbiAgfVxufSJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8cGFyZW50LWNvbXBvbmVudD48L3BhcmVudC1jb21wb25lbnQ-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d

### Workaround

Workaround is to not use a `class` with a variable on the `table` itself, but on e.g. a `div` wrapping the `table`.

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

2.6.1

### Browser/OS/Node environment

Browser: Chrome 109.0.5414.87
OS: macOS Monterey 12.6.3

The issue is reproducible on the lit.dev Playground, too.

**Comment by justinfagnani** [comment]
This is due to the HTML parser, as specified in the HTML spec. `<table>` elements can _only_ have a few elements as children (`<tr>`, `<thead>`, etc.) See https://html.spec.whatwg.org/multipage/tables.html#the-table-element

Any element not allowed in a `<table>` is moved to _before_ the table element.

So `<table><div></div></table>` parses as `<div></div><table></table>`.

Since lit-html passes your template strings to `innerHTML` to create the HTML `<template>` element, we are subject to the browsers rules. The elements are moved without our knowledge. And with bindings, it's really likely that the movement messes up our binding location logic which assumes that binding order and DOM order match up.

So basically... you can't do this. If you want to put custom elements in tables they have to be where all HTML content is permitted, like `<td>` elements.  If you want some helpers to create rows and such, I'd recommend using lit-html _templates_ directly, instead of components, like:

```ts
@customElement("parent-component")
export class ParentComponent extends LitElement {
  render() {
    return html`<table class="my-class">
      ${renderTableBody({myProperty: "my value"})}
    </table>`;
  }
}
```

**Comment by Pephers** [comment]
Thanks @justinfagnani. Coincidentally, at around the same time you posted your comment, I actually realized that the web component is indeed being placed outside of the `table` which probably explains the (for me) cryptic exception and why it cannot render.

Thanks for your suggestion, that will work just as well for my purposes. Perhaps I need to switch to a more "partial templates"-mindset. I'm coming from a React background where components aren't handled by the DOM itself.
